File: demo_visual/benchmark_small.py

import torch
import time

def main():
    m = 64
    k = 784
    n = 128
    print(f"Benchmarking PyTorch Small MatMul [{m}, {k}] x [{k}, {n}] ...")
    
    a = torch.ones(m, k)
    b = torch.ones(k, n)
    
    # Warmup
    _ = torch.matmul(a, b)
    
    start = time.time()
    iters = 1000
    for _ in range(iters):
        _ = torch.matmul(a, b)
    end = time.time()
    
    print(f"PyTorch Average Time: {(end - start) / iters:.6f} s")

if __name__ == "__main__":
    main()

File: demo_visual/src/bin/benchmark_small.rs
use rustorch_core::Tensor;
use std::time::Instant;

fn main() {
    let m = 64;
    let k = 784;
    let n = 128;
    println!(
        "Benchmarking Small MatMul [{}, {}] x [{}, {}] ...",
        m, k, k, n
    );

    let a = Tensor::ones(&[m, k]);
    let b = Tensor::ones(&[k, n]);

    // Warmup
    let _ = a.matmul(&b);

    let start = Instant::now();
    let iters = 1000;
    for _ in 0..iters {
        let _ = a.matmul(&b);
    }
    let duration = start.elapsed();

    println!(
        "Rust Average Time: {:.6} s",
        duration.as_secs_f32() / iters as f32
    );
}
